fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -0,0 +1,383 @@
+
+    /**
+     * Lists all references in the subspace. Puts all references in the cache.
+     * Logs the count of references, and the total size of the data.
+     * All references are put into the cache because each composite file will be required to serve a query and the file references are small.
+     *
+     * @return String list of names of lucene file references
+     */
+    @Override
+    public String[] listAll() {
+        List<String> outList = new ArrayList<>();
+        List<String> displayList = new ArrayList<>();
+        long totalSize = 0L;
+        for (KeyValue kv : txn.getRange(metaSubspace.range())) {
+            String name = metaSubspace.unpack(kv.getKey()).getString(0);
+            outList.add(name);
+            FDBLuceneFileReference fileReference = new FDBLuceneFileReference(Tuple.fromBytes(kv.getValue()));
+            if (name.endsWith(".cfs") || name.endsWith(".si") || name.endsWith(".cfe")) {
+                try {
+                    seekData(name, CompletableFuture.completedFuture(fileReference), 0);
+                } catch (IOException ioe) {
+                    LOG.warn("Cannot Prefetch resource={}", name);
+                    LOG.warn("Prefetch Error", ioe);
+                }
+            }
+            this.fileReferenceCache.put(name, fileReference);
+            if (LOG.isDebugEnabled()) {
+                if (kv.getValue() != null) {
+                    displayList.add(name + "(" + fileReference.getSize() + ")");
+                    totalSize += fileReference.getSize();
+                }
+            }
+        }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("listAllFiles -> count={}, totalSize={}", outList.size(), totalSize);
+            LOG.debug("Files -> {}", displayList);
+        }
+
+        return outList.toArray(new String[0]);
+    }
+
+    /**
+     * deletes the file reference under the provided name.
+     * @param name the name for the file reference
+     */
+    @Override
+    public void deleteFile(String name) throws IOException {
+        try {
+            LOG.trace("deleteFile -> {}", name);
+            hasWritten = true;
+            getFDBLuceneFileReference(name).thenAcceptAsync(
+                    (value) -> {
+                        if (value == null) {
+                            throw new RuntimeException(new NoSuchFileException(name));
+                        }
+                        txn.clear(metaSubspace.pack(name));
+                        txn.clear(dataSubspace.subspace(Tuple.from(value.getId())).range());
+                        this.fileReferenceCache.invalidate(name);
+                    }
+            ).join();
+        } catch (CompletionException completionException) {
+            if (completionException.getCause() instanceof RuntimeException &&
+                    completionException.getCause().getCause() instanceof NoSuchFileException) {
+                throw (NoSuchFileException) completionException.getCause().getCause();
```

Comment:
nitpick here: It might be prettier if you threw the RuntimeException instead of the CompletionException here

Reply:
I was trying to match the characteristics for existing Lucene implementations.  Added a comment to clarify.